## CCAA Python Training
<p> Jupyter notebook for Python Training. </p> 
<p> Date Created: September 8th 2021 </p> 
<p> Written By: Analyst 134 </p>
<p> Data Used: Denver Crimes Last 5 Years 2015 - 2020 </p> 
<p> Data Source: <a href="https://www.denvergov.org/opendata/dataset/city-and-county-of-denver-crime">Denver Crimes </a> </p> 

### Step 1:Import Libraries and Data

In [1]:
#Import Libraries 
import pandas as pd
import numpy as np 
import os 
import geopandas as gpd
import calendar

In [2]:
# Import data 
file = r'C:\Users\rstevenson\Desktop\Desktop Folders\DHSEM\Python Scripts\Data for Python\Denver_Crimes.csv'
Denver_Crimes = pd.read_csv(file)
Denver_Crimes.head()

INCIDENT_ID    OFFENSE_ID  OFFENSE_CODE  OFFENSE_CODE_EXTENSION  \
0  2.018870e+09  2.020000e+15          2399                       0   
1  2.015664e+09  2.020000e+15          5441                       0   
2  2.017601e+10  2.020000e+16          2399                       1   
3  2.019601e+10  2.020000e+16          2308                       0   
4  2.018862e+09  2.020000e+15          5016                       0   

                  OFFENSE_TYPE_ID OFFENSE_CATEGORY_ID FIRST_OCCURRENCE_DATE  \
0                     theft-other             larceny      12/27/2018 15:58   
1                traffic-accident    traffic-accident       11/13/2015 7:45   
2                   theft-bicycle             larceny        6/8/2017 13:15   
3                 theft-from-bldg             larceny       12/7/2019 13:07   
4  violation-of-restraining-order    all-other-crimes      12/22/2018 20:15   

  LAST_OCCURRENCE_DATE     REPORTED_DATE         INCIDENT_ADDRESS      GEO_X  \
0                  NaN  12/27/2018 16:51        2681 N HANOVER CT  3178210.0   
1                  NaN   11/13/2015 8:38  4100 BLOCK W COLFAX AVE  3129148.0   
2       6/8/2017 17:15    6/12/2017 8:44             1705 17TH ST  3140790.0   
3      12/7/2019 18:30   12/9/2019 13:35         1350 N IRVING ST  3132400.0   
4     12/22/2018 20:31  12/22/2018 22:00      13625 E RANDOLPH PL  3188580.0   

       GEO_Y     GEO_LON    GEO_LAT  DISTRICT_ID  PRECINCT_ID NEIGHBORHOOD_ID  \
0  1700715.0 -104.866156  39.755561          5.0        512.0       stapleton   
1  1694748.0 -105.040760  39.739991          1.0        122.0     west-colfax   
2  1699792.0 -104.999264  39.753669          6.0        612.0   union-station   
3  1694088.0 -105.029208  39.738134          1.0        122.0     west-colfax   
4  1716158.0 -104.828868  39.797750          5.0        521.0       montbello   

   IS_CRIME  IS_TRAFFIC  
0         1           0  
1         0           1  
2         1           0  
3         1           0  
4         1           0

In [3]:
# Get the shape of your data set 
Denver_Crimes.shape

(538172, 19)

### Step 2: Identify and Remove any extra data
<p> This data set has 538,172 Rows and 19 Columns! Thats a alot,  you wont need all of that, so lets clean it up! </p> 
<p> <a href ="https://www.codegrepper.com/code-examples/python/how+to+extract+year+from+date+in+pandas"> How to Extract Year from date column</a></p>
<p> <a href ="https://stackoverflow.com/questions/13411544/delete-a-column-from-a-pandas-dataframe"> How to Remove Columns</a></p>
<p><a href ="https://stackoverflow.com/questions/25122099/move-column-by-name-to-front-of-table-in-pandas"> How to Move Columns in Pandas </a></p> 
<p><a href = "https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.query.html"> How to Query Columns Pandas Dataframe</a></p>  
<p><a href ="https://stackoverflow.com/questions/31621414/share-data-between-ipython-notebooks/60863662"> How to Use Magic Commands to Store a Variable </a></p> 

In [4]:
# Create a Year column from the Reported Date. 
# This will help us query the data for the year we are interested in. 
Denver_Crimes['year'] = pd.DatetimeIndex(Denver_Crimes['REPORTED_DATE']).year
Denver_Crimes.head()

INCIDENT_ID    OFFENSE_ID  OFFENSE_CODE  OFFENSE_CODE_EXTENSION  \
0  2.018870e+09  2.020000e+15          2399                       0   
1  2.015664e+09  2.020000e+15          5441                       0   
2  2.017601e+10  2.020000e+16          2399                       1   
3  2.019601e+10  2.020000e+16          2308                       0   
4  2.018862e+09  2.020000e+15          5016                       0   

                  OFFENSE_TYPE_ID OFFENSE_CATEGORY_ID FIRST_OCCURRENCE_DATE  \
0                     theft-other             larceny      12/27/2018 15:58   
1                traffic-accident    traffic-accident       11/13/2015 7:45   
2                   theft-bicycle             larceny        6/8/2017 13:15   
3                 theft-from-bldg             larceny       12/7/2019 13:07   
4  violation-of-restraining-order    all-other-crimes      12/22/2018 20:15   

  LAST_OCCURRENCE_DATE     REPORTED_DATE         INCIDENT_ADDRESS      GEO_X  \
0                  NaN  12/27/2018 16:51        2681 N HANOVER CT  3178210.0   
1                  NaN   11/13/2015 8:38  4100 BLOCK W COLFAX AVE  3129148.0   
2       6/8/2017 17:15    6/12/2017 8:44             1705 17TH ST  3140790.0   
3      12/7/2019 18:30   12/9/2019 13:35         1350 N IRVING ST  3132400.0   
4     12/22/2018 20:31  12/22/2018 22:00      13625 E RANDOLPH PL  3188580.0   

       GEO_Y     GEO_LON    GEO_LAT  DISTRICT_ID  PRECINCT_ID NEIGHBORHOOD_ID  \
0  1700715.0 -104.866156  39.755561          5.0        512.0       stapleton   
1  1694748.0 -105.040760  39.739991          1.0        122.0     west-colfax   
2  1699792.0 -104.999264  39.753669          6.0        612.0   union-station   
3  1694088.0 -105.029208  39.738134          1.0        122.0     west-colfax   
4  1716158.0 -104.828868  39.797750          5.0        521.0       montbello   

   IS_CRIME  IS_TRAFFIC  year  
0         1           0  2018  
1         0           1  2015  
2         1           0  2017  
3         1           0  2019  
4         1           0  2018

In [5]:
# Remove columns you wont need. 
# OFFENSE_ID, OFFENSE_CODE 
# Denver_Crimes = Denver_Crimes.drop('OFFENSE_ID', 1) #If we ran this twice, we would get an error
# Denver_Crimes.drop(['OFFENSE_CODE', 'OFFENSE_CODE_EXTENSION'], axis=1, inplace=True)
# Denver_Crimes.shape
Denver_Crimes.head()

INCIDENT_ID    OFFENSE_ID  OFFENSE_CODE  OFFENSE_CODE_EXTENSION  \
0  2.018870e+09  2.020000e+15          2399                       0   
1  2.015664e+09  2.020000e+15          5441                       0   
2  2.017601e+10  2.020000e+16          2399                       1   
3  2.019601e+10  2.020000e+16          2308                       0   
4  2.018862e+09  2.020000e+15          5016                       0   

                  OFFENSE_TYPE_ID OFFENSE_CATEGORY_ID FIRST_OCCURRENCE_DATE  \
0                     theft-other             larceny      12/27/2018 15:58   
1                traffic-accident    traffic-accident       11/13/2015 7:45   
2                   theft-bicycle             larceny        6/8/2017 13:15   
3                 theft-from-bldg             larceny       12/7/2019 13:07   
4  violation-of-restraining-order    all-other-crimes      12/22/2018 20:15   

  LAST_OCCURRENCE_DATE     REPORTED_DATE         INCIDENT_ADDRESS      GEO_X  \
0                  NaN  12/27/2018 16:51        2681 N HANOVER CT  3178210.0   
1                  NaN   11/13/2015 8:38  4100 BLOCK W COLFAX AVE  3129148.0   
2       6/8/2017 17:15    6/12/2017 8:44             1705 17TH ST  3140790.0   
3      12/7/2019 18:30   12/9/2019 13:35         1350 N IRVING ST  3132400.0   
4     12/22/2018 20:31  12/22/2018 22:00      13625 E RANDOLPH PL  3188580.0   

       GEO_Y     GEO_LON    GEO_LAT  DISTRICT_ID  PRECINCT_ID NEIGHBORHOOD_ID  \
0  1700715.0 -104.866156  39.755561          5.0        512.0       stapleton   
1  1694748.0 -105.040760  39.739991          1.0        122.0     west-colfax   
2  1699792.0 -104.999264  39.753669          6.0        612.0   union-station   
3  1694088.0 -105.029208  39.738134          1.0        122.0     west-colfax   
4  1716158.0 -104.828868  39.797750          5.0        521.0       montbello   

   IS_CRIME  IS_TRAFFIC  year  
0         1           0  2018  
1         0           1  2015  
2         1           0  2017  
3         1           0  2019  
4         1           0  2018

In [6]:
# Move the year Column 
# This code moves the column year to the front. 
Denver_Crimes = Denver_Crimes[['year'] + [ col for col in Denver_Crimes.columns if col != 'year']]
Denver_Crimes.head()

year   INCIDENT_ID    OFFENSE_ID  OFFENSE_CODE  OFFENSE_CODE_EXTENSION  \
0  2018  2.018870e+09  2.020000e+15          2399                       0   
1  2015  2.015664e+09  2.020000e+15          5441                       0   
2  2017  2.017601e+10  2.020000e+16          2399                       1   
3  2019  2.019601e+10  2.020000e+16          2308                       0   
4  2018  2.018862e+09  2.020000e+15          5016                       0   

                  OFFENSE_TYPE_ID OFFENSE_CATEGORY_ID FIRST_OCCURRENCE_DATE  \
0                     theft-other             larceny      12/27/2018 15:58   
1                traffic-accident    traffic-accident       11/13/2015 7:45   
2                   theft-bicycle             larceny        6/8/2017 13:15   
3                 theft-from-bldg             larceny       12/7/2019 13:07   
4  violation-of-restraining-order    all-other-crimes      12/22/2018 20:15   

  LAST_OCCURRENCE_DATE     REPORTED_DATE         INCIDENT_ADDRESS      GEO_X  \
0                  NaN  12/27/2018 16:51        2681 N HANOVER CT  3178210.0   
1                  NaN   11/13/2015 8:38  4100 BLOCK W COLFAX AVE  3129148.0   
2       6/8/2017 17:15    6/12/2017 8:44             1705 17TH ST  3140790.0   
3      12/7/2019 18:30   12/9/2019 13:35         1350 N IRVING ST  3132400.0   
4     12/22/2018 20:31  12/22/2018 22:00      13625 E RANDOLPH PL  3188580.0   

       GEO_Y     GEO_LON    GEO_LAT  DISTRICT_ID  PRECINCT_ID NEIGHBORHOOD_ID  \
0  1700715.0 -104.866156  39.755561          5.0        512.0       stapleton   
1  1694748.0 -105.040760  39.739991          1.0        122.0     west-colfax   
2  1699792.0 -104.999264  39.753669          6.0        612.0   union-station   
3  1694088.0 -105.029208  39.738134          1.0        122.0     west-colfax   
4  1716158.0 -104.828868  39.797750          5.0        521.0       montbello   

   IS_CRIME  IS_TRAFFIC  
0         1           0  
1         0           1  
2         1           0  
3         1           0  
4         1           0

In [7]:
# Create a new data frame to seperate out the data. 
# Query the year column for all crimes in 2020.
Denver_Crimes_2020 = Denver_Crimes[Denver_Crimes.year == 2020]
Denver_Crimes_2020.head()

year   INCIDENT_ID    OFFENSE_ID  OFFENSE_CODE  OFFENSE_CODE_EXTENSION  \
20   2020  2.020132e+09  2.020000e+15          5707                       0   
71   2020  2.020121e+09  2.020000e+15          5441                       0   
99   2020  2.020298e+08  2.020000e+14          2305                       0   
135  2020  2.020431e+08  2.020000e+14          5212                       1   
156  2020  2.020466e+08  2.020000e+14          2303                       0   

                   OFFENSE_TYPE_ID       OFFENSE_CATEGORY_ID  \
20            criminal-trespassing          all-other-crimes   
71                traffic-accident          traffic-accident   
99        theft-items-from-vehicle  theft-from-motor-vehicle   
135  weapon-by-prev-offender-powpo          all-other-crimes   
156                 theft-shoplift                   larceny   

    FIRST_OCCURRENCE_DATE LAST_OCCURRENCE_DATE    REPORTED_DATE  \
20        2/29/2020 20:15                  NaN  2/29/2020 21:16   
71        2/25/2020 11:00                  NaN  2/25/2020 11:02   
99        1/13/2020 23:00       1/14/2020 5:00  1/14/2020 11:07   
135        1/20/2020 2:02                  NaN   1/20/2020 3:36   
156       1/21/2020 15:43                  NaN  1/21/2020 17:13   

           INCIDENT_ADDRESS      GEO_X      GEO_Y     GEO_LON    GEO_LAT  \
20         2222 LAWRENCE ST  3143852.0  1700184.0 -104.988366  39.754698   
71   E 9TH AVE / N OGDEN ST  3147586.0  1691372.0 -104.975266  39.730449   
99    1900 BLK W PACIFIC PL  3137804.0  1672896.0 -105.010399  39.679879   
135  2100 BLOCK ARAPAHOE ST  3143688.0  1699548.0 -104.988962  39.752955   
156  8298 E NORTHFIELD BLVD  3170908.0  1711073.0 -104.891880  39.784132   

     DISTRICT_ID  PRECINCT_ID            NEIGHBORHOOD_ID  IS_CRIME  IS_TRAFFIC  
20           6.0        611.0                five-points         1           0  
71           6.0        623.0               capitol-hill         0           1  
99           4.0        422.0  college-view-south-platte         1           0  
135          6.0        611.0                five-points         1           0  
156          5.0        511.0                  stapleton         1           0

In [16]:
# Convert Reported Date to Date type to remove the time stamps that you dont need. 
# this will be helpful when we create our heat map calendar. 
Denver_Crimes_2020['timestamp'] = pd.to_datetime(Denver_Crimes_2020['REPORTED_DATE'])
Denver_Crimes_2020['New_Date'] = Denver_Crimes_2020['timestamp'].dt.date
Denver_Crimes_2020['Months'] =  pd.to_datetime(Denver_Crimes_2020['New_Date']).dt.month
Denver_Crimes_2020.head()

year   INCIDENT_ID    OFFENSE_ID  OFFENSE_CODE  OFFENSE_CODE_EXTENSION  \
0  2020  2.020132e+09  2.020000e+15          5707                       0   
1  2020  2.020121e+09  2.020000e+15          5441                       0   
2  2020  2.020298e+08  2.020000e+14          2305                       0   
3  2020  2.020431e+08  2.020000e+14          5212                       1   
4  2020  2.020466e+08  2.020000e+14          2303                       0   

                 OFFENSE_TYPE_ID       OFFENSE_CATEGORY_ID  \
0           criminal-trespassing          all-other-crimes   
1               traffic-accident          traffic-accident   
2       theft-items-from-vehicle  theft-from-motor-vehicle   
3  weapon-by-prev-offender-powpo          all-other-crimes   
4                 theft-shoplift                   larceny   

  FIRST_OCCURRENCE_DATE LAST_OCCURRENCE_DATE    REPORTED_DATE  ...    GEO_LAT  \
0       2/29/2020 20:15                  NaN  2/29/2020 21:16  ...  39.754698   
1       2/25/2020 11:00                  NaN  2/25/2020 11:02  ...  39.730449   
2       1/13/2020 23:00       1/14/2020 5:00  1/14/2020 11:07  ...  39.679879   
3        1/20/2020 2:02                  NaN   1/20/2020 3:36  ...  39.752955   
4       1/21/2020 15:43                  NaN  1/21/2020 17:13  ...  39.784132   

   DISTRICT_ID  PRECINCT_ID            NEIGHBORHOOD_ID  IS_CRIME  IS_TRAFFIC  \
0          6.0        611.0                five-points         1           0   
1          6.0        623.0               capitol-hill         0           1   
2          4.0        422.0  college-view-south-platte         1           0   
3          6.0        611.0                five-points         1           0   
4          5.0        511.0                  stapleton         1           0   

     New_Date Months  monthName           timestamp  
0  2020-02-29      2   February 2020-02-29 21:16:00  
1  2020-02-25      2   February 2020-02-25 11:02:00  
2  2020-01-14      1    January 2020-01-14 11:07:00  
3  2020-01-20      1    January 2020-01-20 03:36:00  
4  2020-01-21      1    January 2020-01-21 17:13:00  

[5 rows x 24 columns]

In [17]:
Denver_Crimes_2020['monthName'] = Denver_Crimes_2020['Months'].apply(lambda x: calendar.month_name[x])
Denver_Crimes_2020

year   INCIDENT_ID    OFFENSE_ID  OFFENSE_CODE  OFFENSE_CODE_EXTENSION  \
0      2020  2.020132e+09  2.020000e+15          5707                       0   
1      2020  2.020121e+09  2.020000e+15          5441                       0   
2      2020  2.020298e+08  2.020000e+14          2305                       0   
3      2020  2.020431e+08  2.020000e+14          5212                       1   
4      2020  2.020466e+08  2.020000e+14          2303                       0   
...     ...           ...           ...           ...                     ...   
82424  2020  2.020586e+09  2.020000e+15          1313                       0   
82425  2020  2.020550e+09  2.020000e+15          1315                       1   
82426  2020  2.020601e+10  2.020000e+16          2304                       0   
82427  2020  2.020582e+09  2.020000e+15          5707                       0   
82428  2020  2.020601e+10  2.020000e+16          2305                       0   

                     OFFENSE_TYPE_ID           OFFENSE_CATEGORY_ID  \
0               criminal-trespassing              all-other-crimes   
1                   traffic-accident              traffic-accident   
2           theft-items-from-vehicle      theft-from-motor-vehicle   
3      weapon-by-prev-offender-powpo              all-other-crimes   
4                     theft-shoplift                       larceny   
...                              ...                           ...   
82424                 assault-simple  other-crimes-against-persons   
82425          aggravated-assault-dv            aggravated-assault   
82426       theft-parts-from-vehicle      theft-from-motor-vehicle   
82427           criminal-trespassing              all-other-crimes   
82428       theft-items-from-vehicle      theft-from-motor-vehicle   

      FIRST_OCCURRENCE_DATE LAST_OCCURRENCE_DATE    REPORTED_DATE  ...  \
0           2/29/2020 20:15                  NaN  2/29/2020 21:16  ...   
1           2/25/2020 11:00                  NaN  2/25/2020 11:02  ...   
2           1/13/2020 23:00       1/14/2020 5:00  1/14/2020 11:07  ...   
3            1/20/2020 2:02                  NaN   1/20/2020 3:36  ...   
4           1/21/2020 15:43                  NaN  1/21/2020 17:13  ...   
...                     ...                  ...              ...  ...   
82424       9/30/2020 10:00                  NaN  9/30/2020 11:26  ...   
82425        9/12/2020 7:34                  NaN   9/12/2020 7:34  ...   
82426       9/12/2020 14:45      9/12/2020 14:45  9/12/2020 14:57  ...   
82427       9/28/2020 17:59                  NaN  9/28/2020 17:59  ...   
82428        9/26/2020 4:00       9/26/2020 5:00  9/26/2020 16:56  ...   

         GEO_LAT  DISTRICT_ID  PRECINCT_ID            NEIGHBORHOOD_ID  \
0      39.754698          6.0        611.0                five-points   
1      39.730449          6.0        623.0               capitol-hill   
2      39.679879          4.0        422.0  college-view-south-platte   
3      39.752955          6.0        611.0                five-points   
4      39.784132          5.0        511.0                  stapleton   
...          ...          ...          ...                        ...   
82424  39.761350          5.0        512.0               central-park   
82425  39.733895          1.0        122.0                 villa-park   
82426  39.756793          2.0        211.0                five-points   
82427  39.741300          1.0        123.0                    auraria   
82428  39.728107          3.0        321.0                lowry-field   

       IS_CRIME  IS_TRAFFIC    New_Date Months  monthName           timestamp  
0             1           0  2020-02-29      2   February 2020-02-29 21:16:00  
1             0           1  2020-02-25      2   February 2020-02-25 11:02:00  
2             1           0  2020-01-14      1    January 2020-01-14 11:07:00  
3             1           0  2020-01-20      1    January 2020-01-20 03:36:00  
4             1           0

In [19]:
# Remove the timestamp column
Denver_Crimes_2020 = Denver_Crimes_2020.drop(columns=['timestamp'])

# Get new size of new data frame 
Denver_Crimes_2020.shape

(82429, 23)

In [14]:
#Reindex the dataframe 
Denver_Crimes_2020 = Denver_Crimes_2020.reset_index(drop=True)
Denver_Crimes_2020.head()

year   INCIDENT_ID    OFFENSE_ID  OFFENSE_CODE  OFFENSE_CODE_EXTENSION  \
0  2020  2.020132e+09  2.020000e+15          5707                       0   
1  2020  2.020121e+09  2.020000e+15          5441                       0   
2  2020  2.020298e+08  2.020000e+14          2305                       0   
3  2020  2.020431e+08  2.020000e+14          5212                       1   
4  2020  2.020466e+08  2.020000e+14          2303                       0   

                 OFFENSE_TYPE_ID       OFFENSE_CATEGORY_ID  \
0           criminal-trespassing          all-other-crimes   
1               traffic-accident          traffic-accident   
2       theft-items-from-vehicle  theft-from-motor-vehicle   
3  weapon-by-prev-offender-powpo          all-other-crimes   
4                 theft-shoplift                   larceny   

  FIRST_OCCURRENCE_DATE LAST_OCCURRENCE_DATE    REPORTED_DATE  ...  \
0       2/29/2020 20:15                  NaN  2/29/2020 21:16  ...   
1       2/25/2020 11:00                  NaN  2/25/2020 11:02  ...   
2       1/13/2020 23:00       1/14/2020 5:00  1/14/2020 11:07  ...   
3        1/20/2020 2:02                  NaN   1/20/2020 3:36  ...   
4       1/21/2020 15:43                  NaN  1/21/2020 17:13  ...   

      GEO_LON    GEO_LAT  DISTRICT_ID  PRECINCT_ID            NEIGHBORHOOD_ID  \
0 -104.988366  39.754698          6.0        611.0                five-points   
1 -104.975266  39.730449          6.0        623.0               capitol-hill   
2 -105.010399  39.679879          4.0        422.0  college-view-south-platte   
3 -104.988962  39.752955          6.0        611.0                five-points   
4 -104.891880  39.784132          5.0        511.0                  stapleton   

   IS_CRIME  IS_TRAFFIC    New_Date  Months  monthName  
0         1           0  2020-02-29       2   February  
1         0           1  2020-02-25       2   February  
2         1           0  2020-01-14       1    January  
3         1           0  2020-01-20       1    January  
4         1           0  2020-01-21       1    January  

[5 rows x 23 columns]

In [15]:
# Store your variables using magic commands
%store Denver_Crimes_2020

Stored 'Denver_Crimes_2020' (DataFrame)
